sampler.go

```diff
@@ -286,33 +328,30 @@ func (s *adaptiveSampler) Close() {
 }
 
 func (s *adaptiveSampler) Equal(other Sampler) bool {
-	if o, ok := other.(*adaptiveSampler); ok {
-		if !mapKeysEqual(s.samplers, o.samplers) {
-			return false
-		}
-		if s.lowerBound != o.lowerBound || s.defaultSamplingProbability != o.defaultSamplingProbability {
-			return false
-		}
-		for operation, sampler := range s.samplers {
-			if !sampler.Equal(o.samplers[operation]) {
-				return false
-			}
-		}
-		return true
-	}
 	return false
 }
 
-func mapKeysEqual(m1, m2 map[string]Sampler) bool {
-	if len(m1) != len(m2) {
-		return false
-	}
-	for k := range m1 {
-		if _, ok := m2[k]; !ok {
-			return false
+func (s *adaptiveSampler) update(
+	lowerBound, defaultSamplingProbability float64,
+	samplingRates map[string]float64,
+) error {
+	// TODO handle case where an operation is not present in the new samplingRates, but exists in the sampler
```

Comment:
I'll have a separate map that keeps hold of newly seen operations with a TTL. If the newly seen operation makes it through the pipeline and ends up included in the update call, it will be removed from the map. Otherwise, the TTL will take care of the operation. This way, if the newly seen operation doesn't make it through the pipeline before the next update, it isn't deleted.

This might be overly complicating things and we could always delete any entry that doesn't exist and take the cost of 2~3x lower bound traces while the operation goes through the pipeline.